components/Home/HomeDownloadButton.tsx

```diff
@@ -1,41 +1,38 @@
 import LocalizedLink from '../LocalizedLink';
 import { useNextraContext } from '../../hooks/useNextraContext';
+import { useNodeReleases } from '../../hooks/useNodeReleases';
 import { getNodejsChangelog } from '../../util/getNodeJsChangelog';
-import type { NodeVersionData } from '../../types';
 import type { FC } from 'react';
 
-type HomeDownloadButtonProps = Pick<
-  NodeVersionData,
-  'isLts' | 'node' | 'nodeMajor' | 'nodeNumeric'
->;
+type HomeDownloadButtonProps = {
+  releaseType: 'lts' | 'current';
+};
 
-const HomeDownloadButton: FC<HomeDownloadButtonProps> = ({
-  node,
-  nodeMajor,
-  nodeNumeric,
-  isLts,
-}) => {
+const HomeDownloadButton: FC<HomeDownloadButtonProps> = ({ releaseType }) => {
   const {
     frontMatter: { labels },
   } = useNextraContext();
+  const release = useNodeReleases()[releaseType];
```

Comment:
```suggestion
  const { versionWithPrefix, major, version, isLts } = useNodeReleases()[releaseType];
```

Can we adopt this way of destructuring props from hooks on all usages of `useNodeReleases`?

I know the type is "optional" but I wonder if the `lts, current` will ever be null. Let's change the types of this Hook by adding a type guard.

Reply:
For example (util):

```tsx
export const isNodeRelease = (release: any): release is NodeRelease => typeof release === 'object' && release?.version
```

Reply:
Allowing something of sort:

```tsx
// Maybe add that children needs to have a { release } prop on it at least
type ChildrenType = React.FC<{ release: NodeRelease }>;

type WithNodeReleaseProps = { releaseStatus: ReleaseStatus, children: ChildrenType };

export const WithNodeRelease: FC<WithNodeReleaseProps> = ({ releaseStatus, children: Component }) => {
  const { getReleaseByStatus } = useNodeReleases();

  return useMemo(() => {
    const release = getReleaseByStatus(releaseStatus);

    if (isNodeRelease(release) {
      return <Component release={release} />;
    }

    return null;
  }, [releaseStatus, children]
}
```

This is just one way (by creating a portal) that ensures that release is always defined or the component does not render. It is a practical an clean approach of sorts.

Reply:
Meaning:

```tsx
<WithNodeRelease releaseStatus="Active LTS">
  {HomeDownloadButton}
</WithNodeRelease>
```

Reply:
This allows use to remove the `null` check and directly from `HomeDownloadButton` have the props being a NodeRelease, which allows us to destructure the props directly as `{ versionWithPrefix, isLts, major, version ... }`

Also it acts a "guard" (not just a type guard).

Reply:
I agree, I think that doing `if` check for each time we are calling the hook is too repetitive. But I couldn't figure out the proper approach.

Having a wrapper like you described is clean I think. Thank you!

But, my immediate thought to this: how can we pass the props without making TS upset?
I've never done something like this before. Let me try playing around with it first.

Reply:
Following the approach I mentioned should make TS happy. Or is it complaining about something?

Reply:
Sorry, I phrased it badly.

Tried it, following your approach is okay.

```js
<WithNodeRelease releaseStatus="Active LTS">
  {HomeDownloadButton}
</WithNodeRelease>
```

But if `HomeDownloadButton` has other Props, how can we pass them?

Reply:
It should not have other props.

Reply:
The reason I am asking is because:
We have some components in nodejs.dev that require other Props (e.g. [`DownloadAdditional`](https://github.com/nodejs/nodejs.dev/blob/main/src/components/DownloadComponents/DownloadAdditional/index.tsx)).

If it can't receive other Props then, we can't reuse the Wrapper for these kind of cases which means different components can have different implementations.

Reply:
```tsx
<WithNodeRelease releaseStatus="Active LTS">
  {release=> <DownloadAdditional extraProp="something" release={release}} />
</WithNodeRelease>
```

Reply:
```tsx
<WithNodeRelease releaseStatus="Active LTS">
  {release=> <DownloadAdditional extraProp="something" release={release} />}
</WithNodeRelease>
```

Reply:
@HinataKah0 ☝️

Reply:
Got it!
My understanding was bad 🤣 
Now I finally understand how it works, thanks!

Reply:
After playing around, I omitted the Component memoization and memoize only `getReleaseByStatus` instead.

The reason was:
Let's defer the Component memoization to itself (because [`React.memo`](https://react.dev/reference/react/memo) exists). Not to mention that re-rendering doesn't seem to be triggered often.